### PR TITLE
gh-96663: Add a better error message for __dict__-less classes setattr

### PR DESCRIPTION
**Pull request by Gobot1234**
I think this addresses all the issues I have with the current message. Thanks to Eryk for the pointer as to where I should be editing.

* Issue: gh-96663

### CODE REVIEW

File: Objects/object.c [hunk 1]
```diff
@@ -1577,9 +1579,11 @@ _PyObject_GenericSetAttrWithDict(PyObject *obj, PyObject *name,
         }
         else {
             PyErr_Format(PyExc_AttributeError,
-                         "'%.100s' object has no attribute '%U'",
+                         "'%.100s' object has no attribute '%U' and no "
+                         "__dict__ for setting new attributes",
```
Comment (Gobot1234) [inline Objects/object.c]: I can't find the tests for this bit of code, are there any?

Comment (thatbirdguythatuknownot) [inline Objects/object.c]: This path *may* have `__dict__` available. The only example I can think of that gets to this path with `__dict__` available is this:
```py
import _thread
a = _thread._local()
print(a.__dict__) # {}
del a.b # AttributeError!
```
(from [`local_setattro` in _threadmodule.c](https://github.com/python/cpython/blob/main/Modules/_threadmodule.c#L954))

**Review by thatbirdguythatuknownot — COMMENTED** [review]
There's an example that raises the error message when a `__dict__` is available, so there has to be a check in the second path.

File: Objects/object.c (hunk 1 above)
Comment (Gobot1234) [reply Objects/object.c]: Why would that case go through setattr though?

Comment (thatbirdguythatuknownot) [reply Objects/object.c]: > Why would that case go through setattr though?

@Gobot1234 `setattro` handles both deletion (`value` is `NULL`) and setting of attributes (`value` is an object). It's also handled as a case in [`_PyObject_GenericSetAttrWithDict`](https://github.com/python/cpython/blob/main/Objects/object.c#L1565-L1568):
```c
if (value == NULL)
    res = PyDict_DelItem(dict, name);
else
    res = PyDict_SetItem(dict, name, value);
```

**Comment by Gobot1234** [comment]
I'm not entirely sure why the tests are failing 
```pycon
>>> class B:
...     y = 0
...     __slots__ = ('z', 'foo')
... 
>>> B().fod = 1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'B' object has no attribute 'fod' and no __dict__ for setting new attributes. Did you mean: 'foo'?
```
AFAICT this should be the error that you get? Unless it's failing due to not being printed.

**Comment by thatbirdguythatuknownot** [comment]
> I'm not entirely sure why the tests are failing
> 
> ```
> >>> class B:
> ...     y = 0
> ...     __slots__ = ('z', 'foo')
> ... 
> >>> B().fod = 1
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> AttributeError: 'B' object has no attribute 'fod' and no __dict__ for setting new attributes. Did you mean: 'foo'?
> ```
> 
> AFAICT this should be the error that you get? Unless it's failing due to not being printed.

@Gobot1234 The suggestion (`Did you mean: 'foo'?`) does not seem to be part of the error.

**Comment by ionite34** [comment]
Should this perhaps check for `__setattr__` overrides? There can be cases where the class has no `__dict__` but some non slot attribute setting can still happen. In such a case this error might be misleading.

```py
class Foo:
	__slots__ = ("x", "y")
	
	def __setattr__(self, name, val):
		if name == "all":
			self.x = val
			self.y = val
		else:
			super().__setattr__(name, val)
	
f = Foo()
f.all = 5
```

File: Lib/test/test_class.py [hunk 2]
```diff
@@ -2,7 +2,6 @@
 
 import unittest
 
-
```
Comment (thatbirdguythatuknownot) [inline Lib/test/test_class.py]: why remove this newline?

File: Lib/test/test_class.py [hunk 3]
```diff
@@ -653,7 +660,15 @@ class B:
...
+        with self.assertRaisesRegex(
+            AttributeError,
+            "'B' object has no attribute 'x' and no __dict__ for setting new attributes"
+        ):
+            B().x = 0
+        with self.assertRaisesRegex(
+            AttributeError,
+            "'C' object has no attribute 'x'"
+        ):
             B().x = 0
```
Comment (thatbirdguythatuknownot) [inline Lib/test/test_class.py]:
```suggestion
            C().x = 0
```

**Comment by Gobot1234** [comment]
@pablogsal since this is an error message change would you be willing to review?

File: Lib/test/test_descrtut.py [hunk 4]
```diff
@@ -8,9 +8,9 @@
 # of much interest anymore), and a few were fiddled to make the output
 # deterministic.
 
-from test.support import sortdict
 import doctest
 import unittest
+from test.support import sortdict
```
Comment (Gobot1234) [inline Lib/test/test_descrtut.py]:
```suggestion
from test.support import sortdict
import doctest
import unittest
```

**Review by ethanfurman — APPROVED** [review]

**Review by gvanrossum — APPROVED** [review]

**Pull request merged**